elkscmd/man/man8/login.8
.TH login 8
.SH NAME
login \- log on to the system
.SH SYNOPSIS
.B login
[username]
.SH DESCRIPTION
.B login
is used at the beginning of each terminal session to verify that the users
password is entered corrently,
change to the correct user id and group id, initialise
the users environment, change to the users home directory and start the users
shell.
.SH OPTIONS
.TP 15
.I "[username]"
Log in the user with the given name. If none is specified then a prompt
is printed and a username accepted.
.SH ENVIRONMENT
.B login
does not query the environment, but sets up the following variables:\-
.TP
.SB USER
Set to the users username.
.TP
.SB HOME
Set to the users home directory.
.TP
.SB SHELL
Set to the users shell.
.TP
.SB TERM
Set to the current terminal.
.SH EXIT STATUS
.TP
.I 1
Failed to log in user.
.SH FILES
.PD 0
.TP
.B /etc/passwd
.TP
.B /var/utmp
.PD
.SH BUGS
The
.SB TERM 
variable is always set to "ansi".
.SH AUTHORS
Al Riddoch (ajr@ecs.soton.ac.uk)
.br
David Murn (scuffer@hups.apana.org.au)
.br
Shane Kerr (kerr@wizard.net)
.SH SEE ALSO
.BR getty (8),
.BR init (8).


elks/CHANGELOG
This file is used to record changes made to the CVS database. Please add to
the top of this file before committing any changes. Please use the header
format provided for each entry, noting that the text on each line should not
be longer than the lines of dashes.

Al Riddoch.

Contributor           Email address                     CHANGELOG entries found
~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~
Alan Cox              alan@redhat.com                   17 Jun 2001-18 Jun 2001
Alastair Bridgewater  nyef@softhome.net                 19 Nov 1998- 6 Dec 1998
Alistair Riddoch      alriddoch@zepler.org              25 Sep 1998-17 Jun 2001
Arnaud Launay         asl@launay.org                    15 Jul 1999-17 Oct 2001
Christian Mardmuller  chm@kdt.de                        10 Nov 1999-29 Feb 2000
Christoph Plattner    Christoph.Plattner@dot.at         31 Jan 2000- 2 Feb 2000
Claudio Matsuoka      claudio@conectiva.com
Daniele Giacomini     daniele@swlibero.org		20 Sep 2004
David Murn            scuffer@hups.apana.org.au         11 Nov 1998-15 Nov 1999
Eduardo Habkost	      ehabkost@conectiva.com.br		01 Jun 2004
Giles Russell                                           22 Nov 1999
Greg Haerr                                              14 Jul 1999-20 Jul 1999
Harry Kalogirou       HarKal@gmx.net                    20 Jun 2001- 2 Jan 2003
Jacek Lipkowski       sq5bpf@acid.ch.pw.edu.pl          20 Jan 2002
Jay Barr                                                27 Oct 2002
Ken Martwick          kenm@efn.org                      12 May 2002
M Warner Losh         imp@village.org                   20 Jan 2002
Manuel Novoa III      mjn3@codepoet.org                 18 Jun 2002-25 Jun 2002
Mario Frasca                                            20 Jan 2000-14 Feb 2000
Miguel Bolanos	      mike@linuxlabs.com		20 Sep 2004
Paul Nasrat           pauln@truemesh.com                29 Oct 2001-12 Jul 2002
Per Olofsson          MagerValp@cling.gu.se             25 Feb 2000
Peter L Vachuska      pvachusk@nconnect.net             13 Aug 2001
Richard Wallman       murkygoth@murkygoth.uklinux.net    2 Apr 2002
Riley Williams        Riley@Williams.Name               16 Nov 1998-13 Jul 2003
Roman Festchook       roma@polesye.net                  27 Oct 2001
Simon Weijgers        simon@weijgers.com                16 Sep 1998-15 Nov 1999
Simon Wood                                               6 Jan 2000-12 Jan 2000
Thomas McWilliams     tgm9999@yahoo.com                 27 Aug 1999- 2 Apr 2001
Tommy McCabe 	      rocketjet314@yahoo.com		31 May 2004

-------------------------------------------------------------------------------

Thu Feb 16 17:03:00 EST 2012	Jody Bruchon <jody@jodybruchon.com>	
	* Removed elksfs. It was a copy of the MINIX fs and no tools
	  exist to make such a filesystem.

Thu Nov 23 21:35:28 EST 2006	Segin <segin2005@gmail.com>
	* Changed name for the submenu for selecting charater
	  device drivers to note the fact that that submenu is broken;
	  attempting to enter it will have lxdialog die horribly.

Wed Nov  8 17:49:52 EST 2006	Segin <segin2005@gmail.com>

	* Changed ELKS init sequence to check all locations for 'init'
	  that the Linux kernel does. This is to follow more 
	  "traditional" behaviour.

Sun Aug 20 21:51:25 UTC 2006	Claudio Matsuoka <claudio@mandriva.com>

	* Allow loading Minix executables with total size larger than
	  64KB, such as ash (70140 bytes). Currently the total size
	  minus 32 bytes of header was limited to 16-bit.

	* Set tty1 back to minor 0. This is different from Linux
	  numbering, but it's necessary to have current ELKS virtual
	  terminals working. This and the executable size change put
	  ELKS back in usable state, allowing logins in the first VT
	  and executing a shell.

Mon Sep 20 23:27:23 CST 2004	Daniele Giacomini <daniele@swlibero.org>

	* Added italian keys support, Miguel Bolanos cleaned the original
	  file code.

Wed Jun  9 17:27:52 BRT 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* fix for my last fix: s/rp_cmd/rq_cmd/

Thu Jun 03 21:47:13 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* fixed format specifiers to correctly displaysegment sizes as unsigned
	  values, and to display debug messages correctly.

Thu Jun 03 07:48:25 CST 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* Fixed wrong struct field name for arch/i86/drivers/block/doshd.c
	* Several fixes ro arch/i86/mm/user.c

Web Jun 02 14:16:57 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* Added an extra fix to Eduardos memcpy fix.

Tue Jun 01 21:00:03 CST 2004	Eduardo Habkost ehabkost@conectiva.com.br

	* Fix memcpy_{to,from}fs, adding code to save the contents of SI and DI

Mon May 31 07:39:47 CST 2004	Tommy McCabe <rocketjet314@yahoo.com>

	* Fix wrong assignment (typo?) on printk
	* Fix order of LD parameters on arch/i86/Makefile
	* Use #ifdef rightly on setup.S
	* Fix init to use tty0, not tty1 as system console.

===============================================================================

Sun Jul 13 04:24:44 BST 2003	Riley Williams <Riley@Williams.Name>

	* This is ELKS 0.1.3-pre1

Sat Jul 12 12:24:36 BST 2003	Riley Williams <Riley@Williams.Name>

	* Rewrote kernel/printk.c to allow printk() to correctly print out
	  signed values.
	* Rewrote kernel/printk.c to produce a less verbose panic() dump,
	  and also to increase the number of lines of stack trace to 10.
	* Various spelling corrections to printk() messages.

===============================================================================

Thu Jan 02 16:12:36 EET 2003	Harry Kalogirou <harkal@gmx.net>

        * Better select hash.
	* This is ELKS 0.1.2

-------------------------------------------------------------------------------

Fri Nov 01 01:52:44 EET 2002	Harry Kalogirou <harkal@gmx.net>

        * Changed the position of ds in the _registers struct.
        * Added support for loading the new ELKS executable format. This is 

elkscmd/sys_utils/login.c
/*
 * file:	 login.c
 * descripton:	 login into a user account
 * author:	 Alistair Riddoch <ajr@ecs.soton.ac.uk>
 * modification: David Murn <scuffer@hups.apana.org.au>
 *		 Added password entry
 * modification: Shane Kerr <kerr@wizard.net>
 *		 More work on password entry
 * modification: Al Riddoch <ajr@ecs.soton.ac.uk>  31 Jan 1999
 *		 Added utmp handling, and invocation by getty.
 */

/* todo:  use a non-echoing input routine for password (i.e. getpass) */
/*	  needs a more robust tty first */
/* todo:  add a timeout for serial and network logins */
/*	  need a signal mechanism (i.e. alarm() and SIGALRM) */

#include <unistd.h>
#include <sys/types.h>
#include <pwd.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <errno.h>
#include <dirent.h>
#include <utmp.h>
#include <limits.h>
#include <signal.h>

/*#define USE_UTMP*/	/* Disabled until we fix the "utmp file corrupt" */
			/* issue. 17/4/2002 Harry Kalogirou */

#define PATHLEN 256
#define STR_SIZE (PATHLEN + 7)

void login(register struct passwd * pwd, struct utmp * ut_ent)
{
	int envno = 0;
	char **newenv;
	char user_env[STR_SIZE];
	char shell_env[STR_SIZE];
	char home_env[STR_SIZE];
	char sh_name[STR_SIZE];
	char *renv[16];

	if (fchown(0,pwd->pw_uid,pwd->pw_gid)<0) perror("login (chown)");
#ifdef USE_UTMP
	ut_ent->ut_type = USER_PROCESS;
	strncpy(ut_ent->ut_user, pwd->pw_name, UT_NAMESIZE);
	ut_ent->ut_user[UT_NAMESIZE-1] = '\0';
	time(&ut_ent->ut_time);
	pututline(ut_ent);
	endutent();
#endif
	if (setgid(pwd->pw_gid)<0)
	{
		write(STDOUT_FILENO, "setgid failed\n", 14);
		exit(1);
	}
	if (setuid(pwd->pw_uid)<0)
	{
		write(STDOUT_FILENO, "setuid failed\n", 14);
		exit(1);
	}

	/* copy capitalized /bootopts env vars to login environment */
	newenv = renv;
	while (*environ) {
		if (*environ[0] >= 'A' && *environ[0] <= 'Z')
			newenv[envno++] = *environ;
		environ++;
	}

	strcpy(user_env,"USER=");
	strncpy(user_env+5,pwd->pw_name, STR_SIZE - 6);
	newenv[envno++] = user_env;

	strcpy(home_env,"HOME=");
	strncpy(home_env+5,pwd->pw_dir, STR_SIZE - 6);
	newenv[envno++] = home_env;

	strcpy(shell_env,"SHELL=");
	strncpy(shell_env+6,pwd->pw_shell, STR_SIZE - 7);
	newenv[envno++] = shell_env;
	newenv[envno++] = "TERM=ansi";
	newenv[envno] = NULL;

	*sh_name = '-';
	strncpy(sh_name+1,pwd->pw_shell, STR_SIZE - 1);

	if (chdir(pwd->pw_dir)<0)
		write(STDOUT_FILENO, "No home directory. Starting in /\n", 33);

	environ = renv;
	execl(pwd->pw_shell,sh_name,(char*)0);

	write(STDOUT_FILENO, "No shell (errno ", 16);
	write(STDOUT_FILENO, itoa(errno), 2);
	write(STDOUT_FILENO, ")\n", 2);
	exit(1);
}

int main(int argc, char ** argv)
{
	struct passwd *pwd;
	struct utmp *entryp = NULL;
	char lbuf[64], *pbuf, salt[3];
	char *p;


	signal(SIGTSTP, SIG_IGN);		/* ignore ^Z stop signal*/
	for (;;) {
		if (argc == 1) {
			write(STDOUT_FILENO,"login: ",7);
			errno = 0;
			if (read(STDIN_FILENO, lbuf, sizeof(lbuf)) < 1) {
				if (errno == EINTR)
					continue;
				exit(1);
			}
			p=strchr(lbuf,'\n');
			if (p) *p='\0';
		} else {
			strncpy(lbuf, argv[1], UT_NAMESIZE);
			lbuf[UT_NAMESIZE - 1] = '\0';
			argc = 1;
		}
		pwd = getpwnam(lbuf);
#ifdef USE_UTMP
		struct utmp entry;
		struct utmp newentry;
		char *tty_name;

		if ((tty_name = ttyname(0)) == NULL) {
			goto not_tty;
		}
		entry.ut_type = INIT_PROCESS;
		entry.ut_id[0] = *(tty_name + 8);
		entry.ut_id[1] = *(tty_name + 9);
		if ((entryp = getutid(&entry)) == NULL) {
			write(STDERR_FILENO, "utmp file corrupt\n", 18);
			endutent();
			exit(1);
		}
		newentry = *entryp;
		entryp = &newentry;
		entryp->ut_type = LOGIN_PROCESS;
		strncpy(entryp->ut_user, lbuf, UT_NAMESIZE);
		entryp->ut_user[UT_NAMESIZE-1] = '\0';
		time(&entryp->ut_time);
		pututline(entryp);
		endutent();
not_tty:
#endif
		if ((pwd != NULL) && (pwd->pw_passwd[0] == 0)) {
			login(pwd, entryp);
		}
		pbuf = getpass("Password:");
		write(STDOUT_FILENO, "\n", 1);
		if (pwd != NULL) {
			salt[0]=pwd->pw_passwd[0];
			salt[1]=pwd->pw_passwd[1];
			salt[2]=0;
			if (!strcmp(crypt(pbuf,salt),pwd->pw_passwd)) {
				login(pwd, entryp);
			}
		}
		write(STDOUT_FILENO,"Login incorrect\n\n",17);
	}
}
